File: project/src/app/schemas/inspector.py
from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel, Field


class SiteMapStartRequest(BaseModel):
    max_pages: Optional[int] = Field(None, ge=1, le=50000)
    max_depth: Optional[int] = Field(None, ge=1, le=20)


class SiteMapJobResponse(BaseModel):
    job_id: str
    status: str
    job_type: str
    pages_crawled: int = 0
    pages_discovered: int = 0
    pages_failed: int = 0
    current_depth: int = 0
    started_at: Optional[str] = None
    finished_at: Optional[str] = None
    error: Optional[str] = None


class CrawlEventResponse(BaseModel):
    id: int
    event_type: str
    payload: dict[str, Any]
    created_at: str


class SiteNodeSummary(BaseModel):
    page_key: str
    url: str
    url_pattern: str
    depth: int
    page_type: str
    section: Optional[str]
    title: Optional[str]
    status: str


class SiteTreeNode(BaseModel):
    path: str
    label: str
    page_key: Optional[str] = None
    page_type: Optional[str] = None
    section: Optional[str] = None
    children: list["SiteTreeNode"] = Field(default_factory=list)


SiteTreeNode.model_rebuild()


class SiteGraphResponse(BaseModel):
    nodes: list[SiteNodeSummary]
    edges: list[dict[str, str]]


class SiteSectionResponse(BaseModel):
    section_key: str
    label: str
    root_urls: list[str]
    url_patterns: list[str]
    page_count: int
    useful_score: float


class SitePageDetail(BaseModel):
    page_key: str
    url: str
    url_pattern: str
    depth: int
    parent_page_key: Optional[str]
    page_type: str
    section: Optional[str]
    title: Optional[str]
    excerpt: Optional[str]
    status: str
    meta: Optional[dict[str, Any]]
    outbound_links: list[dict[str, str]] = Field(default_factory=list)

File: frontend/src/api/inspector.ts
import { request } from "./client";

export interface SiteMapJob {
  job_id: string;
  status: string;
  job_type: string;
  pages_crawled: number;
  pages_discovered: number;
  pages_failed: number;
  current_depth: number;
  started_at?: string | null;
  finished_at?: string | null;
  error?: string | null;
}

export interface CrawlEvent {
  id: number;
  event_type: string;
  payload: Record<string, unknown>;
  created_at: string;
}

export interface SiteTreeNode {
  path: string;
  label: string;
  page_key?: string | null;
  url?: string | null;
  page_type?: string | null;
  section?: string | null;
  children: SiteTreeNode[];
}

export interface SiteNodeSummary {
  page_key: string;
  url: string;
  url_pattern: string;
  depth: number;
  page_type: string;
  section?: string | null;
  title?: string | null;
  status: string;
}

export interface SiteGraph {
  nodes: SiteNodeSummary[];
  edges: Array<{ from: string; to: string; type: string }>;
}

export interface SiteMapGroupNode {
  group_key: string;
  parent_group_key?: string | null;
  group_kind: string;
  label: string;
  section?: string | null;
  path_prefix?: string | null;
  url_pattern?: string | null;
  page_type?: string | null;
  page_count: number;
  weight: number;
  inbound_link_count: number;
  representative_page_key?: string | null;
  representative_url?: string | null;
  depth: number;
}

export interface SiteMap {
  nodes: SiteMapGroupNode[];
  edges: Array<{ from: string; to: string; type: string }>;
}

export interface SiteSection {
  section_key: string;
  label: string;
  root_urls: string[];
  url_patterns: string[];
  page_count: number;
  useful_score: number;
}

export interface SitePageDetail {
  page_key: string;
  url: string;
  url_pattern: string;
  depth: number;
  parent_page_key?: string | null;
  page_type: string;
  section?: string | null;
  title?: string | null;
  excerpt?: string | null;
  status: string;
  meta?: Record<string, unknown> | null;
  outbound_links: Array<{ page_key: string; url: string; relation_type: string }>;
}

export const inspectorApi = {
  startSiteMap(maxPages?: number, maxDepth?: number) {
    return request<SiteMapJob>("/api/inspector/site-map/start", {
      method: "POST",
      body: JSON.stringify({ max_pages: maxPages, max_depth: maxDepth }),
    });
  },
  listJobs() {
    return request<SiteMapJob[]>("/api/inspector/jobs");
  },
  getJob(jobId: string) {
    return request<SiteMapJob>(`/api/inspector/jobs/${jobId}`);
  },
  pauseJob(jobId: string) {
    return request<SiteMapJob>(`/api/inspector/jobs/${jobId}/pause`, { method: "POST" });
  },
  resumeJob(jobId: string) {
    return request<SiteMapJob>(`/api/inspector/jobs/${jobId}/resume`, { method: "POST" });
  },
  cancelJob(jobId: string) {
    return request<SiteMapJob>(`/api/inspector/jobs/${jobId}/cancel`, { method: "POST" });
  },
  listEvents(jobId: string, sinceId = 0) {
    return request<CrawlEvent[]>(`/api/inspector/jobs/${jobId}/events?since_id=${sinceId}`);
  },
  getTree(section?: string) {
    const q = section ? `?section=${encodeURIComponent(section)}` : "";
    return request<SiteTreeNode[]>(`/api/inspector/site/tree${q}`);
  },
  getMap(section?: string) {
    const q = section ? `?section=${encodeURIComponent(section)}` : "";
    return request<SiteMap>(`/api/inspector/site/map${q}`);
  },
  getSections() {
    return request<SiteSection[]>("/api/inspector/site/sections");
  },
  getPage(pageKey: string) {
    return request<SitePageDetail>(`/api/inspector/pages/${pageKey}`);
  },
};
